### ISSUE TITLE
[no-floating-promises] Rule not working on async arrow functions

### ISSUE THREAD

**Issue by ATheCoder** [typescript-eslint/typescript-eslint#3358]
- [Y] I have tried restarting my IDE and the issue persists.
- [Y] I have updated to the latest version of the packages.
- [Y] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
module.exports = {
  env: {
    browser: true,
    amd: true,
    node: true,
  },
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: ['tsconfig.json'],
  },
  plugins: ['@typescript-eslint'],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
  ],
  rules: {
    '@typescript-eslint/no-floating-promises': [2]
  }
};

```

```JSON
{
  "include": [
    "src/**/*",
    ".eslintrc.js",
  ],
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "lib": ["es2019"],
    "allowJs": true,
    "outDir": "build",
    "strict": true,
    "noImplicitAny": true,
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "sourceMap": true,
    "typeRoots": ["./node_modules/@types", "./typings"]
  }
}
```

```TS
const a = async () => {console.log('Hello')}
a()
async function foo(){console.log('Bye')}
foo()
```

**Expected Result**

There should be an error when calling both `foo` and `a` without `await`

**Actual Result**

Only calling `foo` without an `await` will cause an error

**Additional Info**

I believe this rule does not work when you assign the result to a variable.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.1` |
| `@typescript-eslint/parser`        | `4.22.1` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.25.0` |
| `node`                             | `14.16.1` |
